Add a doc on optimization techniques for scripts (#4293)
input-output-hk/plutus: doc/reference/optimization.rst

Optimization techniques for Plutus scripts
==========================================

Identifying problem areas
~~~~~~~~~~~~~~~~~~~~~~~~~

In order to identify which parts of the script are responsible for significant resource consumption, you can use the :ref:`profiling support <profiling_scripts>`.

Using strict let-bindings to avoid recomputation

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Perhaps be even more clear that unlike Haskell's let that allows sharing (call-by-need) , our non-strict let do not allow sharing (call-by-name). Hence the recomputations.

.. code-block:: haskell

    -- This may be compiled non-strictly, which could result
    -- in it being evaluated multiple times. However, it will
    -- not be evaluated if it is not used.
    let x = y + z
    in if b then x + x else 1

    -- This will be compiled strictly, but this will mean it
    -- is evaluated even if it is not used.
    let !x = y + z
    in if b then x + x else 1

Specializing higher-order functions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The use of higher-order functions is a common technique to facilitate code reuse.
Higher-order functions are widely used in the Plutus libraries but can be less efficient than specialized versions.

For instance, the Plutus function ``findOwnInput`` makes use of the higher-order function ``find`` to search for the current script input.

.. code-block:: haskell

    findOwnInput :: ScriptContext -> Maybe TxInInfo
    findOwnInput ScriptContext{scriptContextTxInfo=TxInfo{txInfoInputs},
                 scriptContextPurpose=Spending txOutRef} =
        find (\TxInInfo{txInInfoOutRef} -> txInInfoOutRef == txOutRef) txInfoInputs
    findOwnInput _ = Nothing

> This can be rewritten with a recursive function specialized to the specific check in question.

Reviewer: Can rewrite rules help for not having to modify `findOwnInput`? https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/rewrite_rules.html?highlight=specialization%20rules#specialisation

Author: I don't think GHC will do *value* specialization in this way. But anyway, we mostly turn off GHC's optimizer, sadly :(

.. code-block:: haskell

    findOwnInput :: ScriptContext -> Maybe TxInInfo
    findOwnInput ScriptContext{scriptContextTxInfo=TxInfo{txInfoInputs},
                 scriptContextPurpose=Spending txOutRef} = go txInfoInputs
        where
            go [] = Nothing
            go (i@TxInInfo{txInInfoOutRef} : rest) = if txInInfoOutRef == txOutRef
                                                     then Just i
                                                     else go rest
    findOwnInput _ = Nothing

> Common sub-expression elimination

Reviewer 2: ```suggestion
Eliminating common sub-expressions
```

Reviewer 2: ^for more consistent titles.

Author: I mildly prefer this one just because "common subexpression elimination" is a Thing.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When several instances of identical expressions exist within a function’s body, it’s worth replacing them with a single variable to hold the computed value.

In this example, the cost of storing and retrieving ``n * c`` in a single variable is significantly less than recomputing it several times.

.. code-block:: haskell

    let a' = a `divide` n * c
        -- occurrence 1
        b' = b * (n * c)
        -- occurrence 2
        C' = c + (n * c)
    in
      foo a' b' c' n

>     -- only one occurrence

Reviewer: Perhaps mention that as described in `Using strict let-bindings to avoid recomputation` this optimization only works as long as the newly introduced let is going to become a strict let in plutus, thus they probably need to put a bang in there.

    let !t_mul = n * c
        a' = a `divide` t_mul
        b' = b * t_mul
        c' = c + t_mul
    in
      foo a' b' c' n

Using ``error`` for faster failure
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Plutus scripts have access to one impure effect, ``error``, which immediately terminates the script evaluation and will fail validation.
This failure is very fast, but it is also unrecoverable, so only use it in cases where you want to fail the entire validation if there is a failure.

> The Plutus libraries have some functions that fail with ``error``, usually these are given an ``unsafe`` prefix to their name.

Reviewer 3: ```suggestion
The Plutus libraries have some functions that fail with ``error``: usually the names of these have an ``unsafe`` prefix.
```
I feel obliged to say _something_, so here's some minor grammatical quibbling.

Author: Hah, I went further: full stop!
